File: frontend/src/types/intelligence.ts
/**
 * Phase 9 Domain-Specific Clause Intelligence TypeScript Definitions
 */

export interface AlternativePrediction {
  category: string;
  probability: number;
}

export interface RiskAnalysisSummary {
  status: string;
  severity?: string;
  analysis_summary: string;
  risk_details?: string;
}

export interface ClassifyResponse {
  primary_category: string;
  confidence: number;
  classification_source: string;
  model_version?: string;
  dataset_version?: string;
  top_alternatives: AlternativePrediction[];
  explanation_notes?: string;
  risk_analysis: RiskAnalysisSummary;
}

export interface CandidateMetric {
  id: string;
  name: string;
  accuracy: number;
  macro_f1: number;
  weighted_f1: number;
  ece_calibration_error: number;
  latency_ms: number;
  model_size_kb: number;
  delta_f1: number;
}

export interface BenchmarkComparison {
  evaluation_timestamp: string;
  test_samples_evaluated: number;
  candidates: CandidateMetric[];
  selected_model: string;
  selected_model_name: string;
  baseline_macro_f1: number;
  selected_macro_f1: number;
  improvement_delta_f1: number;
  promotion_gate_passed: boolean;
  justification: string;
}

export interface EngineStatus {
  active_engine: string;
  model_name: string;
  model_version: string;
  dataset_version: string;
  fallback_available: boolean;
  benchmark_latency_ms: number;
  baseline_macro_f1: number;
  current_macro_f1: number;
  improvement_delta_f1: number;
  num_categories: number;
  categories: string[];
}

export interface EnhancedClause {
  id: string;
  clauseNumber: string;
  title: string;
  category: string;
  status: string;
  severity?: string;
  obligationType?: string;
  pageNumber: number;
  previewText: string;
  fullExcerpt: string;
  analysisSummary: string;
  riskDetails?: string;
  confidence?: number;
  classificationSource?: "ML_TRANSFORMER" | "DETERMINISTIC_HEURISTIC" | string;
  modelVersion?: string;
  datasetVersion?: string;
  topAlternatives?: AlternativePrediction[];
  explanationNotes?: string;
}

File: backend/app/api/v1/intelligence.py
"""
ClauseGuard Phase 9: Intelligence Engine Endpoints
Provides system diagnostics, benchmark telemetry, and live clause classification API.
"""

import json
from pathlib import Path
from typing import Dict, List, Optional, Any

from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field

from app.intelligence.ml.local_inference_engine import local_intelligence_engine
from app.intelligence.ml.hybrid_classifier import hybrid_clause_classifier, ClassificationResult

router = APIRouter(prefix="/intelligence", tags=["intelligence"])

PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent.parent.parent
BAKEOFF_REPORT_PATH = PROJECT_ROOT / "data" / "models" / "bakeoff_comparison.json"
BASELINE_REPORT_PATH = PROJECT_ROOT / "data" / "models" / "baseline_report.json"
MODEL_PATH = PROJECT_ROOT / "data" / "models" / "production_clause_model.json"


class ClassifyRequest(BaseModel):
    title: str = Field(..., description="Clause title or heading", json_schema_extra={"example": "Clause 7: Handover"})
    text: str = Field(..., description="Clause body text", json_schema_extra={"example": "The Promoter shall deliver physical possession..."})


class AlternativePrediction(BaseModel):
    category: str
    probability: float


class RiskAnalysisSummary(BaseModel):
    status: str
    severity: Optional[str] = None
    analysis_summary: str
    risk_details: Optional[str] = None


class ClassifyResponse(BaseModel):
    primary_category: str
    confidence: float
    classification_source: str
    model_version: Optional[str] = None
    dataset_version: Optional[str] = None
    top_alternatives: List[AlternativePrediction]
    explanation_notes: Optional[str] = None
    risk_analysis: RiskAnalysisSummary


class EngineStatusResponse(BaseModel):
    active_engine: str
    model_name: str
    model_version: str
    dataset_version: str
    fallback_available: bool
    benchmark_latency_ms: float
    baseline_macro_f1: float
    current_macro_f1: float
    improvement_delta_f1: float
    num_categories: int
    categories: List[str]


@router.get("/engine-status", response_model=EngineStatusResponse)
async def get_engine_status():
    """Retrieve operational status, active model version, and benchmark improvement telemetry."""
    baseline_f1 = 0.5720
    current_f1 = 0.8788
    delta_f1 = 0.3068
    latency_ms = 0.030

    if BAKEOFF_REPORT_PATH.exists():
        try:
            with open(BAKEOFF_REPORT_PATH, "r", encoding="utf-8") as f:
                b_data = json.load(f)
                baseline_f1 = b_data.get("baseline_macro_f1", baseline_f1)
                current_f1 = b_data.get("selected_macro_f1", current_f1)
                delta_f1 = b_data.get("improvement_delta_f1", delta_f1)
        except Exception:
            pass

    return EngineStatusResponse(
        active_engine="HYBRID_ML_PROTOTYPE",
        model_name="ClauseGuard Semantic Manifold Prototype Classifier",
        model_version=local_intelligence_engine.model_version,
        dataset_version=local_intelligence_engine.dataset_version,
        fallback_available=True,
        benchmark_latency_ms=latency_ms,
        baseline_macro_f1=baseline_f1,
        current_macro_f1=current_f1,
        improvement_delta_f1=delta_f1,
        num_categories=len(local_intelligence_engine.categories),
        categories=local_intelligence_engine.categories,
    )


@router.get("/benchmarks")
async def get_benchmark_comparison():
    """Retrieve full comparative evaluation bake-off report across all candidate architectures."""
    if not BAKEOFF_REPORT_PATH.exists():
        raise HTTPException(status_code=404, detail="Bake-off benchmark report not found.")
    with open(BAKEOFF_REPORT_PATH, "r", encoding="utf-8") as f:
        return json.load(f)


@router.post("/classify", response_model=ClassifyResponse)
async def classify_clause_text(req: ClassifyRequest):
    """Classify arbitrary clause text using the hybrid intelligence engine."""
    res: ClassificationResult = hybrid_clause_classifier.classify_clause(
        title=req.title, text=req.text
    )

    alts = [
        AlternativePrediction(category=a["category"], probability=a["probability"])
        for a in (res.top_alternatives or [])
    ]

    return ClassifyResponse(
        primary_category=res.category.value if hasattr(res.category, "value") else str(res.category),
        confidence=res.confidence,
        classification_source=res.classification_source,
        model_version=res.model_version,
        dataset_version=res.dataset_version,
        top_alternatives=alts,
        explanation_notes=res.explanation_notes,
        risk_analysis=RiskAnalysisSummary(
            status=res.status,
            severity=res.severity,
            analysis_summary=res.analysis_summary,
            risk_details=res.risk_details,
        ),
    )
